/*
Copyright 2017-2018 Echo Park Labs

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

For additional information, contact:

email: info@echoparklabs.io
*/

syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.epl.protobuf.stac";
option java_outer_classname = "StacMetadata";
option objc_class_prefix = "STC";

package epl.protobuf;

import "google/protobuf/timestamp.proto";
import "epl/protobuf/geometry_operators.proto";

enum RelationshipType {
    FIELD_EQUALS = 0;
    FIELD_NOT_EQUAL = 2;
    FIELD_GREATER_EQUAL = 4;
    FIELD_LESS_EQUAL = 8;
    FIELD_GREATER = 16;
    FIELD_LESS = 32;
    FIELD_RANGE = 64;
    FIELD_NOT_RANGE = 128;
}

enum SortDirection {
    NOT_SORTED = 0;
    DESCENDING = 1;
    ASCENDING = 2;
}

message DoubleField {
    double value = 1;
    RelationshipType rel_type = 2;
    double range_value = 3;
    bool sort_by = 4;
    SortDirection sort_direction = 5;
}

message Int64Field {
    int64 value = 1;
    RelationshipType rel_type = 2;
    int64 range_value = 3;
    bool sort_by = 4;
    SortDirection sort_direction = 5;
}

// TODO make a DateField
message TimestampField {
    google.protobuf.Timestamp value = 1;
    RelationshipType rel_type = 2;
    google.protobuf.Timestamp range_value = 3;
    bool sort_by = 4;
    SortDirection sort_direction = 5;
}

message StringField {
    string value = 1;
    RelationshipType rel_type = 2;
    string range_value = 3;
    bool sort_by = 4;
    SortDirection sort_direction = 5;
}

message GeometryField {
    epl.protobuf.GeometryData geometry = 1;
    epl.protobuf.GeometryBagData geometry_bag = 2;
    epl.protobuf.ServiceOperatorType relational_operator = 3;
    epl.protobuf.OperatorRequest geometry_operator = 4;
}

message BBoxField {
    double upper_left = 1;
    double lower_right = 2;
    double upper_right = 3;
    double lower_left = 4;
}



enum FieldName {
    ALL_FIELD_NAMES = 0;
    EO_CLOUD_COVER = 1;
}


message MetadataRequest {
    uint32 limit = 1;
    uint64 offset = 2;

    SortDirection sort_direction = 3;
    string sort_field = 4;
    repeated string result_fields = 5; // if blank all fields returned

    // TODO really, these should be a float, but python shits the bed at the idea of a 4byte float.
    DoubleField eo_cloud_cover = 6;
    DoubleField eo_sun_azimuth = 7;
    DoubleField eo_sun_elevation = 8;
    DoubleField eo_gsd = 9;
    DoubleField eo_off_nadir = 10;
    // TODO really, these should be a float, but python shits the bed at the idea of a 4byte float.

    BBoxField eo_bbox = 11;
    GeometryField eo_geometry = 12;


    reserved 13 to 200;
    TimestampField src_img_date = 201;
    StringField state_initials = 202;
    StringField filename = 203;


    reserved 500 to max;
}

